pipeline/merger/src/main/scala/uk/ac/wellcome/platform/merger/rules/ItemsRule.scala

```diff
@@ -0,0 +1,145 @@
+package uk.ac.wellcome.platform.merger.rules
+
+import uk.ac.wellcome.models.work.internal.{
+  DigitalLocation,
+  Item,
+  Location,
+  LocationType,
+  TransformedBaseWork,
+  UnidentifiedWork,
+  Unminted
+}
+import uk.ac.wellcome.platform.merger.logging.MergerLogging
+import uk.ac.wellcome.platform.merger.rules.WorkFilters.WorkFilter
+
+/*
+ * Items are merged as follows:
+ *
+ * - The items from digital Sierra works and METS works are added to the items of multi-item
+ *   physical Sierra works, or their locations are added to those of single-item
+ *   physical Sierra works.
+ * - Miro locations are added to single-item Sierra locations.
+ */
+object ItemsRule extends ComposedFieldMergeRule with MergerLogging {
+  type Field = List[Item[Unminted]]
+
+  override def merge(target: UnidentifiedWork,
+                     sources: Seq[TransformedBaseWork]): MergeResult[Field] =
+    MergeResult(
+      field = composeRules(liftIntoTarget)(
+        metsItemsRule,
+        miroItemsRule,
+        physicalDigitalItemsRule)(target, sources).data.items,
```

Comment:
Whenever `composeRules` is used, we just seem to fetch the field in questions data from the output `UnidentifiedWork`, and throw away the rest. Is there a reason we need to lift the data into the target if this is the case?

Reply:
Yeah, unfortunately it is necessary: because we're composing the functions we need to transform the return type into the input type, in this case by lifting the field into the `target`.

We can (and do) keep `sources` constant but in order for this to be a composition the `target` needs to contain the value that the last applied rule produced, rather than that of the 'initial' target.

Reply:
>  the target needs to contain the value that the last applied rule produced, rather than that of the 'initial' target.

Am I right in saying the data for `target` stays the same, apart from the field data that is lifted into it? So a rule could potentially take as arguments the `target` and `sources` (which stay constant), and optionally any latest field data from previously applied rules. That seems to me would avoid having to lift data into the target unnecessarily, but could end up confusing things in other ways, I dunno :man_shrugging:

Reply:
> Am I right in saying the data for target stays the same, apart from the field data that is lifted into it?

Yes

> a rule could potentially take as arguments the target and sources (which stay constant), and optionally any latest field data from previously applied rules

I think this would be more confusing, as the logic related to composition would then have leaked into the rule's implementation. I spent a while playing with rules that took just the field data as arguments, or had access to extractor functions that could get the field data without exposing how this was done, but all of it seemed to introduce more complexity than just working on targets and sources with no extra magic involved.

Reply:
More generally I'm optimistic that composition like this can be done away with if we work out the "true" business logic involved in merging items

Reply:
:+1:

Reply:
@warrd based on the above I stopped using the composition helper in `otherIdentifiers` as it could be done more simply in that specific case - fingers crossed we can do the same for items

Reply:
Yep, I think the original rules are so complicated because the sierra-mets rule, which is the one applied last, had to undo what the previous rules did. I think the actual business logic we want for items is something like:
```
if target has only one item => add mets location if there is a mets work, otherwise add locations from both the digital work and the miro work if they exist
if target has more than one item => add mets item if there is a mets work, otherwise add item from digital work
```

Reply:
@alicefuzier turned the above into code and the tests are still passing... I think it worked?!

Reply:
🎉 🎆